rfcs/20200721-extension-types.md

```diff
@@ -0,0 +1,1304 @@
+encoding.
+
+`StackableTypeSpec` defines the methods `to_boxed_tensor` and
+`from_boxed_tensor` for boxing and unboxing values:
+
+```python
+class StackableTypeSpec(TypeSpec):
+
+  @abstractmethod
+  def to_boxed_tensor(self, value, minimum_rank=0):
+    """Encodes `value` as a stackable Tensor.
+
+    Args:
+      value: A value compatible with this TypeSpec.
+        (Caller is responsible for ensuring compatibility.)
+      minimum_rank: The minimum rank for the returned tensor(s).  This can
+        be used to ensure that the boxed tensor(s) can be unstacked this number
+        of times.
+
+    Return:
+      A `Tensor` (or list of `Tensors`) that encodes `value`.  Stacking, 
+      unstacking, or concatenating boxed tensors must be equivalent to stacking, 
+      unstacking, or concatenating the corresponding unboxed values.
+
+      The returned tensor must have rank greater than or equal to `minimum_rank`.
+
+      If `to_boxed_tensor` returns a list of `Tensors`, then they should be
+      treated as parallel tensors, and corresponding values should be combined.
+      I.e., stacking, unstacking, or concatenating values must be equivalent to
+      stacking, unstacking, or concatenating each of the corresponding tensors
+      from the boxed encoding.  If a list of `Tensors` is returned, they must all
+      have the same shape up to axis `minimum_rank`.
+    """
+
+  @abstractmethod
+  def from_boxed_tensor(self, boxed_tensor):
+  """Decodes `value` from a stackable Tensor.
+
+  Args:
+    boxed_tensor: A `Tensor` (or list of `Tensors`) that was returned by 
+      `to_boxed_tensor`; or a `Tensor` (or list of `Tensors`) that was formed
+      by stacking, unstacking, and concatenating the values returned by
+      `to_boxed_tensor`.
+
+    Returns:
+      A value compatible with this TypeSpec.
+  """
+
+  @abstractmethod
+  def boxed_tensor_spec(self, minimum_rank=0):
+  """Returns a TensorSpec (or list of TensorSpecs) for the boxed tensor encoding.
+
+  Args:
+    minimum_rank: The minimum rank for the returned TensorSpecs.
+
+  Returns:
+    A `TensorSpec` (or list of `TensorSpecs`) that is compatible with 
+    `self.to_boxed_tensor(v, minimum_rank)` for any value `v` that is
+    compatible with this `TypeSpec`.
+  """
+
+  @abstractmethod
+  def stacked(self, num):
+    """Returns a TypeSpec representing stacked objects with this TypeSpec.
+
+    Args:
+      num: An `int` indicating the number of objects that are stacked together,
+        or `None` if the number of objects is not known.
+    """
+
+  @abstractmethod
+  def unstacked(self):
+    """Returns a TypeSpec representing a single unstacked element in this TypeSpec.
+    """
+```
+
+
+**Note:** The `to_boxed_tensor` and `from_boxed_tensor` methods are typically
+implemented by defining new c++ Kernels that encodes values using tensors with
+`dtype=tf.variant`.  The gradient for `to_boxed_tensor` typically calls
```

Comment:
I'm trying to figure out if/how the RFC changes the SavedModel protobuf representation, and how does it affect the downstream compilation stack in the future. 

Does this mean all the extension types will be `DT_VARIENT` in the protobuf representation?

Reply:
Extension types (aka composite tensors) are already handled by SavedModel.  E.g., you can use SavedModel to save/load models that take SparseTensor or RaggedTensor inputs, or return SparseTensor or RaggedTensor outputs. 

In particular, `SavedObjectGraph` contains a set of `SavedConcreteFunction`s.  Those `SavedConcreteFunctions` use `StructuredValue` to record their input & output signatures.  `StructuredValue` (defined in tensorflow/core/protobuf/struct.proto) has a `TypeSpecProto` message that's used to encode `TypeSpecs`.  Currently, it has an enum (`TypeSpecClass`) which indicates the actual type; we will be adding an "EXTENSION_TYPE_SPEC" value to that enum, and a string field to `TypeSpecProto`, which will be used to look up the appropriate TypeSpec in the registry.

When a saved concrete function is called, the `TypeSpec`s that were saved in the `SavedGraphObject` are used to decompose the arguments into component tensors, which are then fed into the saved concrete graph.  If the result is an extension type, then the `TypeSpec` in `SavedConcreteFunction.output_signature` is used to pack the component tensors generated by the saved concrete graph into an appropriate extension type.